stor/dx.py

```diff
@@ -0,0 +1,513 @@
+        """ Glob for pattern relative to this directory."""
+
+        results = list(self.walkfiles(
+            canonicalize=canonicalize,
+            pattern=pattern
+        ))
+        if not results or not results[0]:  # when results == [[]]
+            results = []
+        utils.validate_condition(condition)
+        utils.check_condition(condition, results)
+        return results
+
+    def exists(self):
+        """Checks existence of the path.
+
+        Returns True if the path exists, False otherwise.
+
+        Returns:
+            bool: True if the path exists, False otherwise.
+        """
+        try:
+            # first see if there is a specific corresponding object
+            self.stat()
+            return True
+        except (NotFoundError, ValueError):
+            pass
+        # otherwise we could be a directory, so try to grab first
+        # file/subfolder
+        if self._is_folder():
+            try:
+                self.list(limit=1)
+                return True
+            except NotFoundError:
+                return False
+        return False
+
+    def stat(self):
+        if not self.__stat:
+
+            if self.canonical_resource:
+                self.__stat = dxpy.DXFile(dxid=self.canonical_resource,
+                                          project=self.canonical_project).describe()
+            else:
+                self.__stat = dxpy.DXProject(dxid=self.canonical_project).describe()
+        return self.__stat
+
+
+class DXVirtualPath(DXPath):
+    """Class Handler for DXPath of form 'dx://project-{ID}:/a/b/c' or 'dx://a/b/c'"""
+
+    @property
+    def virtual_project(self):
+        if utils.is_valid_dxid(self.project, 'project'):
+            return dxpy.DXProject(dxid=self.project).name
+        return self.project
+
+    @property
+    def virtual_resource(self):
+        return self.resource
+
+    @property
+    def virtual_path(self):
+        return self
+
+    @cached_property
+    def canonical_project(self):
+        """Returns the unique project that matches the name that user has view access to.
+        If no match is found, returns None
+
+        Raises:
+            DuplicateProjectError - if project name is not unique on DX platform
+            NotFoundError - If project name doesn't exist on DNAnexus
+        """
+        if utils.is_valid_dxid(self.project, 'project'):
+            return self.project
+        else:
+            try:
+                proj_dict = dxpy.find_one_project(
+                    name=self.project, level='VIEW', zero_ok=True, more_ok=False)
+                if proj_dict is None:
+                    raise ProjectNotFoundError('No projects were found with given name ({})'
+                                               .format(self.project))
+                return proj_dict['id']
+            except DXError as e:
+                raise DuplicateProjectError('Duplicate projects were found with given name ({})'
```

Comment:
this seems like a dangerous check, because it's possible you'd swallow up a different error (e.g., about authentication or something else). can you list projects with a limit of 2 and then handle 0/1/2 projects with appropriate exceptions?